solr/core/src/java/org/apache/solr/handler/component/QueryElevationComponent.java

```diff
@@ -376,10 +377,12 @@ protected long getConfigVersion(SolrCore core) {
    *
    * @return The loaded {@link ElevationProvider}; not null.
    */
-  private ElevationProvider loadElevationProvider(SolrCore core) throws Exception {
-    XmlConfigFile cfg;
+  private ElevationProvider loadElevationProvider(SolrCore core) throws IOException, SAXException {
+    InputStream inputStream = null;
+    Document xmlDocument;
     try {
-      cfg = new XmlConfigFile(core.getResourceLoader(), configFileName);
+      inputStream = core.getResourceLoader().openResource(configFileName);
+      xmlDocument = SafeXMLParsing.parseUntrustedXML(log, inputStream);
```

Comment:
I think I completely overlooked something (my bad).  I was initially concerned about using XmlConfigFile because it did not use Solr's SafeXMLParsing utility, and I could also see it enabling some XInclude & entity resolver stuff, which concerned me in its use for elevate.xml.  But this doesn't mean it's not "safe".  Upon further inspection, it appears to be doing some of the same stuff that SafeXMLParsing does in the `parseConfigXML` method, which is relatively safe as it doesn't escape the resource loader.  Do you see @haythemkh ?  If you agree, do you think XmlConfigFile could be modified to call `SafeXMLParsing.parseConfigXml` so that the safe-ness is clear?  If not, we could at least leave a comment there.

At this place in QEC, we could replace these two lines with `SafeXMLParsing.parseConfigXml` and we would in fact continue to support these somewhat exotic XML features in a "safe" way.

Reply:
- About modifying XmlConfigFile with `SafeXMLParsing#parseConfigXML`; It is indeed possible to apply these changes. However, in `XmlConfigFile`, we authorize the non-availability of SystemId so as not to activate xinclude for this purpose. Unlike `SafeXMLParsing#parseConfigXML` where we enable it by default. And, XmlConfigFile is used for reading Solr config files (see `SolrConfig#readXml`). would that have some impact?
- `SafeXMLParsing#parseConfigXML` enables xinclude by default, whereas `SafeXMLParsing#parseUntrustedXML` doesn't. It doesn't seem very "safe", is it? I think if we don't need to activate it we don't have to.

Reply:
@uschindler -- you added the `SafeXMLParsing` library.  My instinct is use use the `parseConfigXML` method for the `elevate.xml` because it conveniently takes a SolrResourceLoader.  It supports x:include too but so-what; right?  It's scoped to the SRL so I don't see how it could be a problem.

Reply:
Yes, i think that's the way to go. parseConfigXML is for stuff that is part of Solr's config. parseUntrustedXML is for stuff coming over the network like a POSTed document.
As all Solr config files supoort xinclude, it should also be supported for elevantion.xml

Reply:
> * About modifying XmlConfigFile with `SafeXMLParsing#parseConfigXML`; It is indeed possible to apply these changes. However, in `XmlConfigFile`, we authorize the non-availability of SystemId so as not to activate xinclude for this purpose. Unlike `SafeXMLParsing#parseConfigXML` where we enable it by default. And, XmlConfigFile is used for reading Solr config files (see `SolrConfig#readXml`). would that have some impact?

We have a valid system id because the ResourceLoader provides one. XInclude will then go through SystemIdResolver provided.

> * `SafeXMLParsing#parseConfigXML` enables xinclude by default, whereas `SafeXMLParsing#parseUntrustedXML` doesn't. It doesn't seem very "safe", is it? I think if we don't need to activate it we don't have to.

If it is a config file, xinclude should be allowed, as this is the standard for SOlr config files. There's also no security risk in this because the xinclude only allows to load files from resource loader and not somewhere outside (it uses SystemIdResolver to resolve includes).

Reply:
Of course if the elevate.xml is coming over the network as part of the request, then it should be parsed with untrusted.

I am not familar with the code, but the rule for config files is:
- if the config file is XML in ResourceLoader, it should use SafeXMLParsing#parseConfigXML with xinclude enabled for consistency
- if the elevation file is passed as part of request body, then parse it using the request input stream using SafeXMLParsing#parseConfigXML

I'd tune XmlConfigFile to use those APIs. Back at that time it was too much work to me, so I left that one out. I just added the SystemIdResolver.

Reply:
Thanks @uschindler for the detailed context!